core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -674,135 +675,232 @@ private RexNode simplifyCoalesce(RexCall call) {
   }
 
   private RexNode simplifyCase(RexCall call, RexUnknownAs unknownAs) {
-    final List<RexNode> operands = call.getOperands();
-    final List<RexNode> newOperands = new ArrayList<>();
-    final Set<String> values = new HashSet<>();
-    for (int i = 0; i < operands.size(); i++) {
-      RexNode operand = operands.get(i);
-      if (RexUtil.isCasePredicate(call, i)) {
-        if (operand.isAlwaysTrue()) {
-          // Predicate is always TRUE. Make value the ELSE and quit.
-          newOperands.add(operands.get(++i));
-          if (unknownAs == FALSE && RexUtil.isNull(operands.get(i))) {
-            values.add(rexBuilder.makeLiteral(false).toString());
-          } else {
-            values.add(operands.get(i).toString());
-          }
-          break;
-        } else if (operand.isAlwaysFalse() || RexUtil.isNull(operand)) {
-          // Predicate is always FALSE or NULL. Skip predicate and value.
-          ++i;
-          continue;
-        }
-      } else {
-        if (unknownAs == FALSE && RexUtil.isNull(operand)) {
-          values.add(rexBuilder.makeLiteral(false).toString());
-        } else {
-          values.add(operand.toString());
-        }
+    List<CaseBranch> branches =
+            CaseBranch.fromCaseOperands(rexBuilder, new ArrayList(call.getOperands()));
+
+    // run simplification on all operands
+    RexSimplify branchSimplifier = this;
+    RelDataType branchType = call.getType();
+
+    for (CaseBranch branch : branches) {
+      // simplify the condition
+      RexNode newCond = branchSimplifier.simplify(branch.cond, RexUnknownAs.FALSE);
+      branch.cond = newCond;
+
+      // use the condition to simplify the branch
+      RexNode value = branch.value;
+      RexNode newValue = branchSimplifier.simplify(value, unknownAs);
+      branch.value = newValue;
+    }
+
+    // remove branches with invalid conditions
+    branches.removeIf(branch -> branch.cond.isAlwaysFalse() || RexUtil.isNull(branch.cond));
```

Comment:
branch is simplified with RexUnknownAs.FALSE, so there's no need to check for isNull case here

Reply:
yes; and I think there are a couple more places - since we are simplifying nulls in simplify() I don't think we need anymore isNull() checks.

I've removed it - and I will probably remove some more from the case related ones.